pkg/exporters/stackdriver/stackdriver_exporter.go

```diff
@@ -0,0 +1,177 @@
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	monitoredres "contrib.go.opencensus.io/exporter/stackdriver/monitoredresource"
+	"github.com/golang/glog"
+	"github.com/spf13/pflag"
+	"go.opencensus.io/stats/view"
+	"google.golang.org/api/option"
+
+	"github.com/avast/retry-go"
+	"k8s.io/node-problem-detector/pkg/exporters"
+	seconfig "k8s.io/node-problem-detector/pkg/exporters/stackdriver/config"
+	"k8s.io/node-problem-detector/pkg/types"
+	"k8s.io/node-problem-detector/pkg/util/metrics"
+)
+
+func init() {
+	clo := commandLineOptions{}
+	exporters.Register(exporterName, types.ExporterHandler{
+		CreateExporterOrDie: NewExporterOrDie,
+		Options:             &clo})
+}
+
+const exporterName = "stackdriver"
+
+var NPDMetricToSDMetric = map[metrics.MetricID]string{
+	metrics.HostUptimeID:      "compute.googleapis.com/guest/system/uptime",
+	metrics.ProblemCounterID:  "compute.googleapis.com/guest/system/problem_count",
+	metrics.DiskAvgQueueLenID: "compute.googleapis.com/guest/disk/queue_length",
+	metrics.DiskIOTimeID:      "compute.googleapis.com/guest/disk/io_time",
+	metrics.DiskWeightedIOID:  "compute.googleapis.com/guest/disk/weighted_io_time",
+}
+
+func getMetricType(view *view.View) string {
+	viewName := view.Measure.Name()
+	// When there is no pre-defined Stackdriver Metric Type, fallback to custom metrics.
+	// e.g. custom.googleapis.com/npd/host/uptime
+	fallbackMetricType := "custom.googleapis.com/npd/" + viewName
+
+	metricID, ok := metrics.MetricMap.ViewNameToMetricID(viewName)
+	if !ok {
+		return fallbackMetricType
+	}
+	stackdriverMetricType, ok := NPDMetricToSDMetric[metricID]
+	if !ok {
+		return fallbackMetricType
+	}
+	return stackdriverMetricType
+}
+
+type stackdriverExporter struct {
+	config seconfig.StackdriverExporterConfig
+}
+
+func (se *stackdriverExporter) setupOpenCensusViewExporterOrDie() {
+	clientOption := option.WithEndpoint(se.config.APIEndpoint)
+
+	var globalLabels stackdriver.Labels
+	globalLabels.Set("instance_name", se.config.Metadata.InstanceName, "The name of the VM instance")
+
+	viewExporter, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID:               se.config.Metadata.ProjectID,
+		MonitoringClientOptions: []option.ClientOption{clientOption},
+		MonitoredResource: &monitoredres.GCEInstance{
+			ProjectID:  se.config.Metadata.ProjectID,
+			InstanceID: se.config.Metadata.InstanceID,
+			Zone:       se.config.Metadata.Zone,
+		},
+		GetMetricType:           getMetricType,
+		DefaultMonitoringLabels: &globalLabels,
+	})
+	if err != nil {
+		glog.Fatalf("Failed to create Stackdriver OpenCensus view exporter: %v", err)
+	}
+
+	exportPeriod, err := time.ParseDuration(se.config.ExportPeriod)
+	if err != nil {
+		glog.Fatalf("Failed to parse ExportPeriod %q: %v", se.config.ExportPeriod, err)
+	}
+
+	view.SetReportingPeriod(exportPeriod)
+	view.RegisterExporter(viewExporter)
+}
+
+func (se *stackdriverExporter) populateMetadataOrDie() {
+	if se.config.Metadata.IsValid() {
+		return
```

Comment:
When it is invalid, why should we just return? Should we die here?

Reply:
:P sorry for the confusing naming. This `if` clause actually triggers when the metadata is valid.

Basically we want this:
1. If the user defined a set of valid metadata in the config file, we want NPD to respect it and use it.
2. If the config file don't have any metadata, or they are not valid, we want NPD to [query GCE metadata server](https://cloud.google.com/compute/docs/storing-retrieving-metadata#querying), and collect them at runtime.

This `if` clause is for use case #1.